client-tez/src/main/java/org/apache/tez/runtime/library/input/RssOrderedGroupedKVInput.java

```diff
@@ -0,0 +1,428 @@
+ * merge them into a single input view.
+ *
+ * The Copy and Merge will be triggered by the initialization - which is handled
+ * by the Tez framework. Input is not consumable until the Copy and Merge are
+ * complete. Methods are provided to check for this, as well as to wait for
+ * completion. Attempting to get a reader on a non-complete input will block.
+ */
+@Public
+public class RssOrderedGroupedKVInput extends AbstractLogicalInput {
+
+  static final Logger LOG = LoggerFactory.getLogger(RssOrderedGroupedKVInput.class);
+
+  protected TezRawKeyValueIterator rawIter = null;
+  protected Configuration conf;
+  protected RssShuffle shuffle;
+  protected MemoryUpdateCallbackHandler memoryUpdateCallbackHandler;
+  private final BlockingQueue<Event> pendingEvents = new LinkedBlockingQueue<>();
+  private long firstEventReceivedTime = -1;
+  @SuppressWarnings("rawtypes")
+  protected ValuesIterator vIter;
+
+  private TezCounter inputKeyCounter;
+  private TezCounter inputValueCounter;
+  private TezCounter shuffledInputs;
+
+  private final AtomicBoolean isStarted = new AtomicBoolean(false);
+
+  public RssOrderedGroupedKVInput(InputContext inputContext, int numPhysicalInputs) {
+    super(inputContext, numPhysicalInputs);
+  }
+
+  @Override
+  public synchronized List<Event> initialize() throws IOException {
+    this.conf = TezUtils.createConfFromUserPayload(getContext().getUserPayload());
+
+    if (this.getNumPhysicalInputs() == 0) {
+      getContext().requestInitialMemory(0L, null);
+      isStarted.set(true);
+      getContext().inputIsReady();
+      LOG.info("input fetch not required since there are 0 physical inputs for input vertex: "
+              + getContext().getSourceVertexName());
+      return Collections.emptyList();
+    }
+
+    long initialMemoryRequest = Shuffle.getInitialMemoryRequirement(conf,
+            getContext().getTotalMemoryAvailableToTask());
+    this.memoryUpdateCallbackHandler = new MemoryUpdateCallbackHandler();
+    getContext().requestInitialMemory(initialMemoryRequest, memoryUpdateCallbackHandler); // TODO
+
+    this.inputKeyCounter = getContext().getCounters().findCounter(TaskCounter.REDUCE_INPUT_GROUPS);
+    this.inputValueCounter = getContext().getCounters().findCounter(
+            TaskCounter.REDUCE_INPUT_RECORDS);
+    this.shuffledInputs = getContext().getCounters().findCounter(
+            TaskCounter.NUM_SHUFFLED_INPUTS);
+    this.conf.setStrings(TezRuntimeFrameworkConfigs.LOCAL_DIRS, getContext().getWorkDirs());
+
+    return Collections.emptyList();
+  }
+
+  @Override
+  public synchronized void start() throws IOException {
+    if (!isStarted.get()) {
+      memoryUpdateCallbackHandler.validateUpdateReceived();
+      // Start the shuffle - copy and merge
+      shuffle = createRssShuffle();
+      shuffle.run();
+      if (LOG.isDebugEnabled()) {
+        LOG.debug("Initialized the handlers in shuffle..Safe to start processing..");
```

Comment:
I can't get the meaning of `Shuffle .. Safe`? Could you explain more?

Reply:
[      LOG.debug("Initialized the handlers in shuffle..Safe to start processing..");](https://github.com/apache/tez/blob/ccad71dabceb850b58ff38e77fd206e68f62ce4b/tez-runtime-library/src/main/java/org/apache/tez/runtime/library/input/OrderedGroupedKVInput.java#L133)

Reply:
`shuffle .. Safe` seems not a correct word. It's also weird grammar. I can't get your thoughts.